tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h

```diff
@@ -354,7 +354,6 @@ void fromMsg(const geometry_msgs::msg::QuaternionStamped& msg, geometry_msgs::ms
  * \param in A instance of the tf2::Quaternion specialization of the tf2::Stamped template.
  * \return The QuaternionStamped converted to a geometry_msgs QuaternionStamped message type.
  */
-template <>
 inline
 geometry_msgs::msg::QuaternionStamped toMsg(const tf2::Stamped<tf2::Quaternion>& in)
```

Comment:
Template specialization allows different return types for the same argument (though the template arguments must be given). I don't think these two could be in the same header as overloads, and maybe not even the same binary if the compiler ignores `inline`.

```c++
geometry::msgs::Quaternion toMsg(const tf2::Stamped<tf2::Quaternion>&);
geometry::msgs::QuaternionStamped toMsg(const tf2::Stamped<tf2::Quaternion>&)
```

That said, I didn't find anyone defining `QuaternionStamped` -> `Quaternion`.

Reply:
I would argue that the definition

```c++
geometry::msgs::Quaternion toMsg(const tf2::Stamped<tf2::Quaternion>&);
```

shouldn't be allowed since there's a loss of information, but perhaps there's valid use-cases for this.

IIUC, if we want to keep the template specialization, for this case to work with `tf2::convert` we should update [this line](https://github.com/ros2/geometry2/blob/71b054249cda9f1bbdbdc2f61bd518b998ce0378/tf2/include/tf2/impl/convert.h#L65) to:

```diff
-  b = toMsg(a);
+  b = toMsg<A, B>(a);
```

Reply:
> I would argue that the definition shouldn't be allowed since there's a loss of information

Maybe so. I'm not sure what the motivation for template specialization is, but it seems like that my example is not a supported use case. If `class Q Stamped` converts to either `msg Q Stamped` or `msg Q`, and `msg Q Stamped` can be converted to `class Q Stamped`, then there would be a path to convert `msg Q Stamped` to `msg Q` via `msg Q Stamped -(fromMsg)> class Q Stamped -(toMsg)> msg Q`, but message to message conversions are forbidden by this comment.

https://github.com/ros2/geometry2/blob/71b054249cda9f1bbdbdc2f61bd518b998ce0378/tf2/include/tf2/impl/convert.h#L43-L45